sympy/functions/combinatorial/factorials.py

```diff
@@ -165,7 +165,7 @@ def eval(cls, n):
 
                     # GMPY factorial is faster, use it when available
                     elif HAS_GMPY:
-                        from sympy.core.compatibility import gmpy
+                        from sympy.external.gmpy import gmpy
                         result = gmpy.fac(n)
```

Comment:
Ideally anything `HAS_GMPY` logic should be handled in the `gmpy` compatibility module. For example there could just be
```
from sympy.external.gmpy import factorial
```
at the top of this file.

Reply:
I see the point and that could work here and for bincoeff, but e.g. primetest would require quite a bit of rewrite in a not obvious way (or one could refactor, call the `gmpy` compatibility module, which call primetest again.

I'd prefer to open an issue for these things as there are probably more than those I have touched upon here. I assume that `HAS_GMPY` really should never be used in the code base?

Reply:
I would prefer for all of the `HAS_GMPY` logic to be as contained as possible but yes that can be a separate issue/PR.

> I assume that `HAS_GMPY` really should never be used in the code base?

Ideally but there doesn't need to be a hard rule. It's just better if you can see in one place what gmpy is doing and reason about what happens when installing it.

I would like to add support for using python_flint instead of gmpy2 because it's faster for many things and also does much more that is useful for sympy:
https://fredrikj.net/python-flint/

Example:
```
In [8]: import flint

In [9]: p = Poly(random_poly(x, 200, 0, 2))

In [10]: coeffs = [int(c) for c in p.all_coeffs()[::-1]]

In [11]: pf = flint.fmpq_poly(coeffs)

In [12]: %time ok = (pf ** 2).factor()
CPU times: user 45.3 ms, sys: 899 µs, total: 46.2 ms
Wall time: 48.5 ms

In [13]: %time ok = (p ** 2).factor_list()
CPU times: user 8.33 s, sys: 93 ms, total: 8.43 s
Wall time: 8.55 s
```
It will be a lot easier to control making use of external dependencies if all the code for doing it is as concentrated as possible.

Also we can avoid duplication and make better use of external libraries better by e.g. using `MPQ` internally as part of Rational.